ir/src/constraint_builder/mod.rs

```diff
@@ -93,30 +77,38 @@ impl ConstraintBuilder {
                 }
 
                 // add the trace access at the specified boundary to the graph.
-                let lhs = self.constraints.insert_trace_access(
-                    &self.symbol_table,
-                    &trace_access,
-                    domain,
-                )?;
+                let lhs = self
+                    .constraints
+                    .insert_trace_access(&self.symbol_table, &trace_access)?;
+                // get the trace segment and domain of the boundary column access
+                let (lhs_segment, lhs_domain) = self.constraints.node_details(&lhs, domain)?;
 
                 // add its expression to the constraints graph.
-                let rhs = self.constraints.insert_expr(
-                    &self.symbol_table,
-                    constraint.value(),
-                    &mut self.variable_roots,
-                    domain,
-                )?;
-
-                // ensure that the inferred trace segment of the rhs expression can be applied to
-                // column against which the boundary constraint is applied.
-                // trace segment inference defaults to the lowest segment (the main trace) and is
-                // adjusted according to the use of random values and trace columns.
-                if lhs.trace_segment() < rhs.trace_segment() {
-                    return Err(SemanticError::trace_segment_mismatch(lhs.trace_segment()));
+                let rhs =
+                    self.constraints
+                        .insert_expr(&self.symbol_table, constraint.value(), domain)?;
+                // get the trace segment and domain of the expression
+                let (rhs_segment, rhs_domain) = self.constraints.node_details(&rhs, domain)?;
+
+                // ensure that the inferred trace segment and domain of the rhs expression can be
+                // applied to column against which the boundary constraint is applied.
+                if lhs_segment < rhs_segment {
+                    // trace segment inference defaults to the lowest segment (the main trace) and is
+                    // adjusted according to the use of random values and trace columns.
+                    return Err(SemanticError::trace_segment_mismatch(lhs_segment));
+                }
```

Comment:
This could happen if we enforce a constraint against the auxiliary trace that uses an element from the main trace.

For example:

```
trace_columns:
    main: [a, b]
    aux: [c]

boundary_constraints:
    c = a + b
```

Let me know if we will never care about this case

Reply:
(We get 12 failing tests when I change this to `!=`)

Reply:
hmmm - such constraints shouldn't be valid as they will never be satisfied.

I think a general rule we can impose (in a different PR) is that mixed addition or subtraction of elements (i.e., when one operand is in the base field and another one is in the extension field) should not be allowed. We can only multiply them (and the result is an extension field element).

Reply:
Ah yeah makes sense that in this example we end up saying an extension field element equals a base field element, which can't be valid. 

To clarify - do you want me to change to `!=` now and address all failing tests now before merging this or should we address this after the refactor when we add the general rule you mentioned

Reply:
Actually, I'm just going to merge this so we can close this refactor off. I'll open a separate issue for this